### PR TITLE
feat(router): add pathParamsChange mode for runGuardsAndResolvers

### PR DESCRIPTION
**Pull request by jasonaden**
This option means guards and resolvers will ignore changes to optional
parameters such as query and matrix params. When the path or any path
params change, guards and resolvers will be run.

Related to discussion in #18253
FW-560 #resolve

### CODE REVIEW

**Comment by jasonaden** [comment]
[Presubmit](http://test/OCL:219365606:BASE:219365738:1540934709066:8120530)

**Review by benlesh — APPROVED** [review]

File: tools/public_api_guard/router/router.d.ts [hunk 1]
```diff
@@ -451,7 +451,7 @@ export declare class RoutesRecognized extends RouterEvent {
     toString(): string;
 }
 
-export declare type RunGuardsAndResolvers = 'paramsChange' | 'paramsOrQueryParamsChange' | 'always';
+export declare type RunGuardsAndResolvers = 'paramsChange' | 'paramsOrQueryParamsChange' | 'always' | 'initialNavOnly';
```
[the remark this answers is not in the document]
Comment (IgorMinar) [reply tools/public_api_guard/router/router.d.ts]: "init" or "ignoreParamChange"? "once"?

**Review by IgorMinar — CHANGES_REQUESTED** [review]

**Comment by benlesh** [comment]
This PR fixes 237 tests in `core:test`.

```
master
1264 specs, 776 failures, 2 pending specs

this PR
1264 specs, 539 failures, 2 pending specs
```

**Comment by JoostK** [comment]
> This PR fixes 237 tests in `core:test`.
> 
> ```
> master
> 1264 specs, 776 failures, 2 pending specs
> 
> this PR
> 1264 specs, 539 failures, 2 pending specs
> ```

@benlesh How? Did your comment end up with the correct PR?

**Comment by jasonaden** [comment]
Updated this PR to better align with how guards and resolvers should be used. The option changed to `pathParamsChange`, meaning that guards and resolvers will be run when either the path or any path params change. This will effectively ignore all *optional* parameters including matrix and URL params.

**Comment by jasonaden** [comment]
Updated [Presubmit](http://test/OCL:219365606:BASE:219663330:1541095362541:2c5b70b4)

File: packages/router/src/config.ts [hunk 2]
```diff
@@ -39,6 +39,9 @@ import {UrlSegment, UrlSegmentGroup} from './url_tree';
  * - `runGuardsAndResolvers` defines when guards and resolvers will be run. By default they run only
  *    when the matrix parameters of the route change. When set to `paramsOrQueryParamsChange` they
  *    will also run when query params change. And when set to `always`, they will run every time.
+ *    When set to `pathParamsOnly`, guards and resolvers will run when the path or any path params
```
Comment (sarunint) [inline packages/router/src/config.ts]:
```suggestion
 *    When set to `pathParamsChange`, guards and resolvers will run when the path or any path params
```

Comment (jasonaden) [reply packages/router/src/config.ts]: 👍

File: tools/public_api_guard/router/router.d.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (jasonaden) [reply tools/public_api_guard/router/router.d.ts]: Changed to `pathParamsChange`

File: packages/router/src/config.ts [hunk 3]
```diff
@@ -39,6 +39,9 @@ import {UrlSegment, UrlSegmentGroup} from './url_tree';
  * - `runGuardsAndResolvers` defines when guards and resolvers will be run. By default they run only
  *    when the matrix parameters of the route change. When set to `paramsOrQueryParamsChange` they
  *    will also run when query params change. And when set to `always`, they will run every time.
+ *    When set to `pathParamsChange`, guards and resolvers will run when the path or any path params
```
Comment (jasonaden) [inline packages/router/src/config.ts]: Created nested list now that there's 4 options.

**Review by IgorMinar — APPROVED** [review]
lgtm after docs update. thanks!

**Comment by simeyla** [comment]
@jasonaden Any reason why a `never` option wasn't added too for completeness?

I have child matrix parameters that I use ONLY to preselect an item from a dropdown. It isn't a component, and I don't want it to be a query parameter for readability.

     example.com/support/contact-us/technical    

So I get the dreaded `Outlet is not activated` error if defining a child route with no component.

As you can see I solve the problem of trying to resolve a component by using a custom `runGuardsAndResolvers() => false`. Had there been a `**never**` option - as somebody here suggested then I could have just used that. 

The solution would have also been a lot more discoverable with `never` as an option too:

```
 {
        path: 'contact-us',
        component: ContactUsComponent,

        children: [
            {
                path: 'topics',
                pathMatch: 'full'
            },
            {
                path: ':category',
                runGuardsAndResolvers: () => false
            }
        ]
    }
```

https://stackoverflow.com/a/54875320/16940

**Pull request closed without merging**